### PR TITLE
Load TrackerData lazily

### PR DESCRIPTION
**Pull request by mallexxx**
Task/Issue URL: https://app.asana.com/0/414709148257752/1202407381883565/f
GH Issue URL: https://github.com/duckduckgo/iOS/issues/1200
BSK PR: https://github.com/duckduckgo/BrowserServicesKit/pull/110
macOS PR: https://github.com/more-duckduckgo-org/macos-browser/pull/796
Tech Design URL:
CC: @miasma13 @bwaresiak 

**Description**:
Load TrackerData lazily and asynchronously to avoid JSON Decoding during init leading to app termination at startup

**Steps to test this PR**:
1. Validate Content Blocking Rules load and work without issues

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

### CODE REVIEW

**Comment by bwaresiak** [comment]
@mallexxx If you can make this up to date, I'll find time to look at it.

File: DuckDuckGo/PrivacyProtectionController.swift [hunk 1]
```diff
@@ -66,6 +65,7 @@ class PrivacyProtectionController: ThemableNavigationController {
             PrivacyProtectionErrorController(coder: coder, configuration: self.privacyConfig)
         }) else { return }
         controller.errorText = errorText
+        controller.delegate = self
```
Comment (mallexxx) [inline DuckDuckGo/PrivacyProtectionController.swift:68]: Seems like the delegate setting when a page displays error was missing here, was it intentional?

**Comment by mallexxx** [comment]
TODO: make it explicit in DEBUG

File: DuckDuckGo/TabViewController.swift [hunk 2]
```diff
@@ -317,96 +288,24 @@ class TabViewController: UIViewController {
...
-            textSizeUserScript,
-            findInPageScript,
-            navigatorPatchScript,
-            surrogatesScript,
-            contentBlockerRulesScript,
-            fingerprintScript,
-            faviconScript,
-            fullScreenVideoScript,
-            autofillUserScript,
-            printingUserScript
```
Comment (mallexxx) [inline DuckDuckGo/TabViewController.swift:412]: why was documentScript (DocumentUserScript) missing here? was it intentional, should I remove it from UserScripts.swift?

**Comment by bwaresiak** [comment]
I did some tests and it seems that attribution is not handled correctly - state is inconsistent. 

Scenario:
1. Enter attribution flow for some domain.
2. On vendor site see that attribution works.
3. Trigger content rule list re-application by toggling protection on and off.
4. Error - attribution is not active on the vendor site even tho it should be.

File: DuckDuckGo/TabViewController.swift (hunk 2 above)
Comment (bwaresiak) [reply DuckDuckGo/TabViewController.swift:412]: Looking back on the blame, I think this is iOS 12 leftover and could be removed, probably? :)

**Comment by mallexxx** [comment]
@bwaresiak I updated the Attribution logic to match the latest macOS implementation, now this scenario seems working as it should

File: DuckDuckGo/TabViewController.swift (hunk 2 above)
Comment (mallexxx) [reply DuckDuckGo/TabViewController.swift:412]: ok, made extra clean-up of document.js and LongPressGestureRecognizer

**Comment by bwaresiak** [comment]
Overall changes to architecture look great, but there is one thing that worries me with regard to original objective:

Plan was to move loading of the data to background / non-main thread or at least decouple it from app startup flow and scene initialization. Now, since ContentBlocking has became a singleton (instead of class with static fields initialized on demand) first access to that loads all the data (config, TDS) which is in fact in app delegate didFinishLaunchingWithOptions (through `PrivacyFeatures.httpsUpgrade.loadDataAsync()`). So this has been moved from Scene Initialization to App Initialization flow - which is not necessarily a bad thing as we should have a bit more resources available there (I think watchdog timeout is larger there) but at the same time this may be something we may want to improve further.

Note that I don't expect this to negatively impact startup performance - after all we have been already do this exact work, just from the MainVC - so we just moved this to earlier point of time.

I'm going to retest content blocking stack Tomorrow and if everything will be fine, let's go with it, unless @miasma13 or @mallexxx  has something to add to my comment. :)

**Comment by mallexxx** [comment]
I‘ve performed launch measurements and the ContentBlocking initialization at init seems totally irrelevant in launch time, StorageCache initialization has much greater effect (caused mainly by TLD init because of ineffective JSONDecoder memory usage) although not really bad
Screenshot 2022-11-03 at 13 26 28

File: DuckDuckGo/ContentBlockingUpdating.swift [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        // 1. Collect updates from ContentBlockerRulesManager and generate UserScripts based on its output
+        cancellable = contentBlockerRulesManager.updatesPublisher
+            // regenerate UserScripts on:
+            // prefs changes notifications with initially published value for combineLatest to work
+            .combineLatest(onNotificationWithInitial(PreserveLogins.Notifications.loginDetectionStateChanged), combine)
+            .combineLatest(onNotificationWithInitial(AppUserDefaults.Notifications.doNotSellStatusChange), combine)
+            .combineLatest(onNotificationWithInitial(AppUserDefaults.Notifications.autofillEnabledChange), combine)
+            .combineLatest(onNotificationWithInitial(AppUserDefaults.Notifications.textSizeChange), combine)
+            .combineLatest(onNotificationWithInitial(AppUserDefaults.Notifications.didVerifyInternalUser), combine)
+            .combineLatest(onNotificationWithInitial(StorageCacheProvider.didUpdateStorageCacheNotification)
```
Comment (bwaresiak) [inline DuckDuckGo/ContentBlockingUpdating.swift:88]: Wasn't that removed from list of notifications that trigger reload?

**Review by bwaresiak — APPROVED** [review]
Great job, one comment in line, one more here, but good to go when addressed:

When I create new tab, reload is being called through combine (TabVC:2124) for that new tab - is that needed?

**Comment by diegoreymendez** [comment]
I just restored this branch to investigate a finding page regression.

**Pull request merged**